### ISSUE TITLE
Run `link-check` sooner, and irrespective of other pipelines

### ISSUE THREAD

**Issue by fredrik-bakke** [UniMath/agda-unimath#775]
It would be nice of `link-check` is ran before type-checking, or at least before the second type-checking stage, so that we don't have to wait 15 minutes to know if our links are broken. Also, it should be ran irrespective of the status of any of the other pipelines.

**Comment by fredrik-bakke** [comment]
Would it be possible to run it on a second kernel?

**Comment by EgbertRijke** [comment]
This is true, but part of the work flow should be to build the website locally before committing. Maybe we should have a separate `make verify` and `make check`, where make check also checks the links

**Comment by VojtechStep** [comment]
The reason why `lint-check` isn't done before typechecking is because it needs to be run on the markdown generated by Agda, so that always needs to run beforehand. Running the markdown generation requires Agda to scope-check the code, and I don't know if the proper typechecking can reuse that work - on the other hand, when we typecheck first, then the scope-checking is effectively skipped, since the whole library is already compiled. I would find it wasteful to run an additional parallel job just for scopechecking and link checking, because it takes a non-trivial amount of resources to replicate work that needs to be done anyway.
The current CI setup caches the typechecking results even if the linkcheck fails, so if a build fails because of incorrect links, the next CI run should be much faster (the run where linkcheck failed took >8 minutes in total, the next one took 3 minutes).

**Comment by VojtechStep** [comment]
I guess we could try to sidestep Agda completely for linkchecking purposes, since we only want to run Agda to get the right file structure, not syntax highlighting. I'll see if I can instruct Agda to do that, but I sincerely doubt it, so we might have to resort to some hack like replicating Agda's logic and copying the files over by ourself in a script.

**Comment by fredrik-bakke** [comment]
This does not seem worthwhile to me. However, isn't it still possible to perform the link-check after the first type-checking step instead of the second?

**Comment by fredrik-bakke** [comment]
Also, I just noticed, both type-checking steps produce highlighting information. That seems unnecessary...

**Comment by EgbertRijke** [comment]
A more sensible setup would be if we'd have

1. `make pre-commit` that does the pre-commit checks and does not type-check the library
2. `make verify` that type-checks the library
3. `make link-check` that attempts to verify links in the library, but does not type-check the library
4. `make check` that does all of the above in that order.

**Comment by VojtechStep** [comment]
> Also, I just noticed, both type-checking steps produce highlighting information. That seems unnecessary...

If I didn't mess something up, and if I'm reading the action output correctly, only the Ubuntu runner produces the highlighted markdown. Could you elaborate?

**Comment by fredrik-bakke** [comment]
Oh, my bad, I must have mixed things up. Indeed, it looks like that step is skipped in the pipeline.
